### PR TITLE
Account for relocs when improving index choice for waterfall loops

### PR DESCRIPTION
**Pull request by dstutt**
Introduction of relocs meant that the code to improve waterfall loop indices no
longer worked. This meant it would always fall back to the value tagged as
nonUniform (usually a resource descriptor).

The pre-existing code to attempt to improve the index to the offset used in the
GEP for the resource descriptor load (it is much smaller) broke because it could
trace back through a reloc and associated code to manipulate it.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit e077bca598591bf7cbec8e1cf18b404d93c346b1
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 0d602bbcfa1a86fd52f66966554b05e384a10d38</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 1598ada40008302abedfd724275329f8e7dea637</li>
</ul></details>

<details><summary>CTS tests (Failed: 3/140859)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24005/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22954/46960 (48.9%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24004/46960 (51.1%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 22955/46960 (48.9%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22646/46939 (48.2%)</li>
<li><details><summary>Failed: 1/46939 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.image_format_properties.3d.optimal.r64_sint
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 24292/46939 (51.8%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
This whole code smells like a re-implementation of divergence analysis...

I suppose it solves an immediate need, but it increasingly seems like we ought to have a lazy variant of divergence analysis :/

**Comment by dstutt** [comment]
Agreed. I debated stripping this out of lgc althogether and re-implementing it as an llvm-ir pass (where DA would be available). However, there are issues with that approach too.
For nearly all cases that lgc will encounter, this code should be fine I think.

Also, this situation shouldn't be encountered in most cases - however, it is required for separate shader compilation.

BTW - I'm going to be putting an llpc-lit test case in for this.

**Comment by dstutt** [comment]
Test added

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0d05d500d4a334e067a391dbe68ea51483016815
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 85f858393412732198fcd3dc0a0efb8653f5d3f9</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev a87d46c4de072f579b9ea51faa6a5814c9a5bfb8</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev f7d29036e33b28778d2fab07cb32d5757c9ad1ee</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

**Comment by dstutt** [comment]
Note: just review the top change "0d05d50"

**Comment by dstutt** [comment]
I think the test failures for this are flakes

**Comment by dstutt** [comment]
Ping

**Review by trenouf — APPROVED** [review]
LGTM

File: lgc/builder/BuilderImpl.cpp [hunk 1]
```diff
@@ -209,14 +229,45 @@ Instruction *BuilderImplBase::createWaterfallLoop(Instruction *nonUniformInst, A
           }
           break;
         }
+        if (auto inttoptr = dyn_cast<IntToPtrInst>(base)) {
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.cpp:232]: inttoptr  -> intToPtr

File: lgc/builder/BuilderImpl.cpp [hunk 2]
```diff
@@ -209,14 +229,45 @@ Instruction *BuilderImplBase::createWaterfallLoop(Instruction *nonUniformInst, A
...
+          base = inttoptr->getOperand(0);
+          continue;
+        }
+        if (auto ptrtoint = dyn_cast<PtrToIntInst>(base)) {
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.cpp:236]: ptrtoint -> ptrToInt

File: lgc/builder/BuilderImpl.cpp [hunk 3]
```diff
@@ -209,14 +229,45 @@ Instruction *BuilderImplBase::createWaterfallLoop(Instruction *nonUniformInst, A
...
+          base = ptrtoint->getOperand(0);
+          continue;
+        }
+        if (auto zext = dyn_cast<ZExtInst>(base)) {
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.cpp:240]: zext -> zExt

File: lgc/builder/BuilderImpl.cpp [hunk 4]
```diff
@@ -209,14 +229,45 @@ Instruction *BuilderImplBase::createWaterfallLoop(Instruction *nonUniformInst, A
...
+                  base = worklist.pop_back_val();
+                  continue;
+                }
                 nonUniformVal = index;
+                break;
+              }
             }
           }
         }
+        if (auto addi = dyn_cast<Instruction>(base)) {
```
Comment (amdrexu) [inline lgc/builder/BuilderImpl.cpp:259]: addi -> addInst

**Review by amdrexu — APPROVED** [review]
Some minor naming issues

File: lgc/builder/BuilderImpl.cpp (hunk 1 above)
Comment (dstutt) [reply lgc/builder/BuilderImpl.cpp]: Done

File: lgc/builder/BuilderImpl.cpp (hunk 2 above)
Comment (dstutt) [reply lgc/builder/BuilderImpl.cpp]: Done

File: lgc/builder/BuilderImpl.cpp (hunk 3 above)
Comment (dstutt) [reply lgc/builder/BuilderImpl.cpp]: Done

File: lgc/builder/BuilderImpl.cpp (hunk 4 above)
Comment (dstutt) [reply lgc/builder/BuilderImpl.cpp]: Done

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8c02d3fd39a77e520803bcaf95dd738968ba585d
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 007816f4bd2fd3ace30f8af9629cff7ec1dcb998</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev b25c1eac19f1773b753ed727314aa3cf2c0897dd</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev e09ed8b4a5ed4201e59b8d5dc66e17dd05eca428</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26164/46960 (55.7%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.synchronization.op.single_queue.barrier.write_image_compute_read_copy_image_to_buffer.image_128x128_r16g16b16a16_uint
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

**Comment by trenouf** [comment]
Single flake. I will merge this.

**Comment by trenouf** [comment]
D'oh -- Jax has already merged it. :-)

**Pull request merged**